sdk/search/search-documents/src/indexModels.ts

```diff
@@ -685,3 +722,136 @@ export type IndexDocumentsAction<T> = {
 } & Partial<T>;
 
 // END manually modified generated interfaces
+
+// Utility types
+
+export type UnionToIntersection<U> =
+  // Distribute members of U into parameter position of a union of functions
+  (
+    U extends unknown ? (_: U) => unknown : never
+  ) extends // Infer the intersection of the members of U as a single intersected parameter type
+  (_: infer I) => unknown
+    ? I
+    : never;
+
+// Types that should not be included in SelectFields recursion
+export type ExcludedODataTypes = Date | GeographyPoint;
+
+/**
+ * Produces a union of valid Cognitive Search OData $select paths for T
+ * using a post-order traversal of the field tree rooted at T.
+ */
+export type SelectFields<T extends object> = T extends Array<infer Elem>
+  ? // Allow selecting fields only from elements which are objects
+    NonNullable<Elem> extends object
+    ? SelectFields<NonNullable<Elem>>
+    : never
+  : {
+      // Only consider string keys
+      [Key in keyof T & string]: NonNullable<T[Key]> extends object
```

Comment:
Thanks for taking a look at this.

I didn't think about preserving the homomorphic mapping, but it would be great if we can make it work. I don't really understand the suggested change, though. If `K in keyof T` is `as`-mapped to `keyof T & string` then doesn't each individual key expand to the set of all keys of T? `[Key in keyof T as Key & string]` (Key vs keyof T on the right-hand side), on the other hand, makes sense to me, still erases non-string keys, and would seem to preserve the homomorphism just as well because the constraint is still a naked keyof.

@dgetu I think this would definitely be nice to have. I agree with the supposition that eventually being able to use refactorings that cross the type-mapping is much more important than preserving the original form of the declaration, but it could also help with optionality in our case.

Reply:
U are right - i’ve written this incorrectly and didnt re-review my own comment. `as Key & string` is the correct thing to do here (i edited my post with the correct suggestion)

Reply:
Thanks for taking a look at this! I think this change does help with making sure the narrowed document type will be assignable to the user-defined model.

It also helps by allowing users to rename their properties and have that propagate to the new type, which is great. I wonder if this feature can be extended to change string literals with a template literal type as well... I'll have to look at the issue tracking this TS feature.

Reply:
> I wonder if this feature can be extended to change string literals with a template literal type as well... I'll have to look at the issue tracking this TS feature.

If I understand correctly your intention - this kind of mapping isn't implemented anywhere in TS atm. When you rename a property using a mapped type then a bunch of functionality is lost. It's definitely doable to improve this situation though (at least for some more straightforward cases). [One of the somewhat related issues](https://github.com/microsoft/TypeScript/issues/46602) is actually added to the 5.1 milestone - so perhaps it will introduce some of the required machinery.

Reply:
That would a truly immense level of analysis, but it would be pretty sick if a type like ``type X<T> = `${keyof T}/test`;`` could refactor a string like `const x: X<{ foo: string }> = "foo/test";` to something else by doing rename-symbol on `foo`. :smile: